File: 백준/Silver/2343. 기타 레슨/기타 레슨.cc
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;

ll N, M;
ll arr[300004];

bool check(ll mid) {
    ll num = 0;
    ll count = 1;
    for (int i = 0; i < N; i++) {
        if (mid >= num + arr[i]) {
            num += arr[i];
        }
        else {
            if (arr[i] > mid) {
                return false;
            }
            num = arr[i];
            count++;
        }
    }
    return M>=count;
}


int main() {
    cin >> N >> M;
    ll l = 1, r = 0;
    for (int i = 0; i < N; i++) {
        cin >> arr[i];
        r += arr[i];;
    }

    ll minResult = 987654321;
    while (l <= r) {
        ll mid = (l + r) / 2;
        if (check(mid)) {
            minResult = min(minResult, mid);
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    cout << minResult;

    return 0;
}

File: 백준/Silver/2343. 기타 레슨/기타 레슨.java
import java.util.Scanner;

public class Main {
    static int N, M;
    static int[] arr = new int[300004];

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        N = scanner.nextInt();
        M = scanner.nextInt();
        
        int l = 1, r = 0;

        for (int i = 0; i < N; i++) {
            arr[i] = scanner.nextInt();
            r += arr[i];
        }

        int minResult = Integer.MAX_VALUE;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (check(mid)) {
                minResult = Math.min(minResult, mid);
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        System.out.println(minResult);

        scanner.close();
    }

    static boolean check(int mid) {
        int num = 0;
        int count = 1;

        for (int i = 0; i < N; i++) {
            if (mid >= num + arr[i]) {
                num += arr[i];
            } else {
                if (arr[i] > mid) {
                    return false;
                }
                num = arr[i];
                count++;
            }
        }

        return M >= count;
    }
}